wagtail/contrib/frontend_cache/utils.py

```diff
@@ -132,36 +132,41 @@ def purge_urls_from_cache(urls, backend_settings=None, backends=None):
             backend.purge_batch(urls)
 
 
-def _get_page_cached_urls(page):
-    page_url = page.full_url
+def _get_page_cached_urls(page, *, request=None):
+    page_url = page.get_full_url(request)
     if page_url is None:  # nothing to be done if the page has no routable URL
         return []
 
-    return [page_url + path.lstrip("/") for path in page.specific.get_cached_paths()]
+    return [
+        page_url + path.lstrip("/")
+        for path in page.specific_deferred.get_cached_paths()
+    ]
 
 
-def purge_page_from_cache(page, backend_settings=None, backends=None):
-    purge_pages_from_cache([page], backend_settings=backend_settings, backends=backends)
+def purge_page_from_cache(page, backend_settings=None, backends=None, *, request=None):
+    urls = _get_page_cached_urls(page, request=request)
+    purge_urls_from_cache(urls, backend_settings, backends)
 
 
-def purge_pages_from_cache(pages, backend_settings=None, backends=None):
-    urls = []
-    for page in pages:
-        urls.extend(_get_page_cached_urls(page))
-
-    if urls:
-        purge_urls_from_cache(urls, backend_settings, backends)
+def purge_pages_from_cache(
+    pages, backend_settings=None, backends=None, *, request=None
+):
+    batch = PurgeBatch(request=request)
+    batch.add_pages(pages)
+    batch.purge(backend_settings, backends)
 
 
 class PurgeBatch:
     """Represents a list of URLs to be purged in a single request"""
 
-    def __init__(self, urls=None):
+    def __init__(self, urls=None, *, request=None):
         self.urls = set()
 
         if urls is not None:
             self.add_urls(urls)
 
+        self.request = request or get_dummy_request()
```

Comment:
Question: Because `get_full_url` is used, could using this request result in the wrong site being used to resolve the URLs hostname?

Reply:
In setups where pages belong to more than one site, it could indeed unexpectedly influence the result of `get_full_url`. The changes in #12454 should mean that we can supply things other than a `HttpRequest` to `get_full_url()` - which would be used for caching, but not for identifying the current site.

Reply:
@RealOrangeOne We are now sending an arbitrary object to `Page.get_full_url()` instead of a `HttpRequest`, so that we're not influencing the result in any way!